File: data-structures/linked-list/linked-list-class/python/linked_list_v1_test.py
import unittest
from node import Node
from linked_list_v1 import LinkedListv1


class LinkedListv1Test(unittest.TestCase):
    def setUp(self):
        self.list = LinkedListv1()

    def test_push_front(self):
        self.list.push_front('A')
        self.assertEqual(self.list.get_size(), 1)
        self.assertEqual(self.list.indexOf('A'), 0)

        self.list.push_front('B')
        self.assertEqual(self.list.get_size(), 2)
        self.assertEqual(self.list.indexOf('B'), 0)
        self.assertEqual(self.list.indexOf('A'), 1)

    def test_push_back(self):
        self.list.push_back('A')
        self.assertEqual(self.list.get_size(), 1)
        self.assertEqual(self.list.indexOf('A'), 0)

        self.list.push_back('B')
        self.assertEqual(self.list.get_size(), 2)
        self.assertEqual(self.list.indexOf('A'), 0)
        self.assertEqual(self.list.indexOf('B'), 1)

    def test_delete_front(self):
        self.list.push_front('A')
        self.list.delete_front()
        self.assertEqual(self.list.get_size(), 0)

        self.list.push_front('A')
        self.list.push_front('B')
        self.list.delete_front()
        self.assertEqual(self.list.get_size(), 1)
        self.assertEqual(self.list.indexOf('B'), -1)
        self.assertEqual(self.list.indexOf('A'), 0)

    def test_delete_back(self):
        self.list.push_back('A')
        self.list.delete_back()
        self.assertEqual(self.list.get_size(), 0)

        self.list.push_back('A')
        self.list.push_back('B')
        self.list.push_back('C')
        self.list.delete_back()
        self.assertEqual(self.list.get_size(), 2)
        self.assertEqual(self.list.indexOf('A'), 0)
        self.assertEqual(self.list.indexOf('B'), 1)
        self.assertEqual(self.list.indexOf('C'), -1)

    def test_contains(self):
        self.assertRaises(ValueError, self.list.contains, 'A')
        self.list.push_front('A')
        self.assertTrue(self.list.contains('A'))
        self.assertFalse(self.list.contains('C'))

    def test_indexOf(self):
        self.assertRaises(ValueError, self.list.indexOf, 'A')
        self.list.push_front('A')
        self.assertEqual(self.list.indexOf('A'), 0)
        self.assertEqual(self.list.indexOf('B'), -1)

    def test_edge_cases(self):
        self.assertRaises(ValueError, self.list.delete_front)
        self.assertRaises(ValueError, self.list.delete_back)
        self.assertRaises(ValueError, self.list.contains, 'A')
        self.assertRaises(ValueError, self.list.indexOf, 'A')

    def tearDown(self):
        self.list = None


if __name__ == '__main__':
    unittest.main()

File: data-structures/linked-list/linked-list-class/java/LinkedListv1Test.java
import java.util.NoSuchElementException;

class LinkedListv1Test {
    public static void main(String[] args) {
        LinkedListv1Test test = new LinkedListv1Test();
        test.runTests();
    }

    public void runTests() {
        testPushFront();
        testPushBack();
        testDeleteFront();
        testDeleteBack();
        testContains();
        testIndexOf();
    }

    public void testPushFront() {
        LinkedListv1 list = new LinkedListv1();
        list.pushFront("C");
        list.pushFront("B");
        list.pushFront("A");

        assert list.getSize() == 3;
        assert list.contains("A");
        assert list.contains("B");
        assert list.contains("C");
        assert list.indexOf("A") == 0;
        assert list.indexOf("B") == 1;
        assert list.indexOf("C") == 2;
    }

    public void testPushBack() {
        LinkedListv1 list = new LinkedListv1();
        list.pushBack("A");
        list.pushBack("B");
        list.pushBack("C");

        assert list.getSize() == 3;
        assert list.contains("A");
        assert list.contains("B");
        assert list.contains("C");
        assert list.indexOf("A") == 0;
        assert list.indexOf("B") == 1;
        assert list.indexOf("C") == 2;
    }

    public void testDeleteFront() {
        LinkedListv1 list = new LinkedListv1();
        list.pushFront("C");
        list.pushFront("B");
        list.pushFront("A");

        list.deleteFront();
        assert list.getSize() == 2;
        assert !list.contains("A");
        assert list.contains("B");
        assert list.contains("C");
        assert list.indexOf("B") == 0;
        assert list.indexOf("C") == 1;

        list.deleteFront();
        assert list.getSize() == 1;
        assert !list.contains("A");
        assert !list.contains("B");
        assert list.contains("C");
        assert list.indexOf("C") == 0;

        list.deleteFront();
        assert list.getSize() == 0;
        assert !list.contains("A");
        assert !list.contains("B");
        assert !list.contains("C");

        try {
            list.deleteFront();
            assert false; // The previous line should throw an exception, so this line should not be
                          // reached
        } catch (NoSuchElementException e) {
            assert true; // An exception should be thrown
        }
    }

    public void testDeleteBack() {
        LinkedListv1 list = new LinkedListv1();
        list.pushBack("A");
        list.pushBack("B");
        list.pushBack("C");

        list.deleteBack();
        assert list.getSize() == 2;
        assert list.contains("A");
        assert list.contains("B");
        assert !list.contains("C");
        assert list.indexOf("A") == 0;
        assert list.indexOf("B") == 1;

        list.deleteBack();
        assert list.getSize() == 1;
        assert list.contains("A");
        assert !list.contains("B");
        assert !list.contains("C");
        assert list.indexOf("A") == 0;

        list.deleteBack();
        assert list.getSize() == 0;
        assert !list.contains("A");
        assert !list.contains("B");
        assert !list.contains("C");

        try {
            list.deleteBack();
            assert false; // The previous line should throw an exception, so this line should not be
                          // reached
        } catch (NoSuchElementException e) {
            assert true; // An exception should be thrown
        }
    }

    public void testContains() {
        LinkedListv1 list = new LinkedListv1();
        list.pushBack("A");
        list.pushBack("B");
        list.pushBack("C");

        assert list.contains("A");
        assert list.contains("B");
        assert list.contains("C");
        assert !list.contains("D");
    }

    public void testIndexOf() {
        LinkedListv1 list = new LinkedListv1();
        list.pushBack("A");
        list.pushBack("B");
        list.pushBack("C");

        assert list.indexOf("A") == 0;
        assert list.indexOf("B") == 1;
        assert list.indexOf("C") == 2;
        assert list.indexOf("D") == -1;
    }
}
